code/modules/events/crystal_event.dm

```diff
@@ -0,0 +1,477 @@
+GLOBAL_LIST_INIT(waves, list("small wave" = list(/obj/structure/crystal_portal/small=4, /obj/structure/crystal_portal/medium=1),
+"medium wave" = list(/obj/structure/crystal_portal/small=4, /obj/structure/crystal_portal/medium=3, /obj/structure/crystal_portal/big=1),
+"big wave" = list(/obj/structure/crystal_portal/small=5, /obj/structure/crystal_portal/medium=3, /obj/structure/crystal_portal/big=2, /obj/structure/crystal_portal/huge=1),
+"huge wave" = list(/obj/structure/crystal_portal/small=7, /obj/structure/crystal_portal/medium=5, /obj/structure/crystal_portal/big=3, /obj/structure/crystal_portal/huge=2)))
+GLOBAL_LIST_EMPTY(crystal_portals)
+GLOBAL_LIST_EMPTY(destabilized_crystals)
+
+/*
+This section is for the event controller
+*/
+/datum/round_event_control/crystal_invasion
+	name = "Crystal Invasion"
+	typepath = /datum/round_event/crystal_invasion
+	weight = 2
+	min_players = 35
+	max_occurrences = 1
+	earliest_start = 45 MINUTES
+
+/datum/round_event/crystal_invasion
+	startWhen = 10
+	announceWhen = 1
+	///Is the name of the wave, used to check wich wave will be generated
+	var/wave_name
+	///Max number of portals that can spawn per type of wave
+	var/portal_numbers
+	///Check if this is the first wave or not
+	var/spawned = FALSE
+
+/datum/round_event/crystal_invasion/start()
+	choose_wave_type()
+
+/datum/round_event/crystal_invasion/announce(fake)
+	priority_announce("WARNING - Destabilization of the Supermatter Crystal Matrix detected, please stand by waiting further instructions", "Alert")
+	sound_to_playing_players('sound/misc/notice1.ogg')
+
+///Choose the type of the wave
+/datum/round_event/crystal_invasion/proc/choose_wave_type()
+	if(!wave_name)
+		wave_name = pickweight(list(
+			"small wave" = 55,
+			"medium wave" = 37,
+			"big wave" = 5,
+			"huge wave" = 3))
+	switch(wave_name)
+		if("small wave")
+			portal_numbers = pick(2, 3, 4, 5)
+		if("medium wave")
+			portal_numbers = pick(4, 5, 6, 7)
+		if("big wave")
+			portal_numbers = pick(6, 7, 8, 9)
+		if("huge wave")
+			portal_numbers = pick(9, 10, 11, 12)
+		else
+			WARNING("Wave name of [wave_name] not recognised.")
+			kill()
+
+	var/list/sm_crystal = list()
+	for(var/obj/machinery/power/supermatter_crystal/temp in GLOB.machines)
+		if(QDELETED(temp))
```

Comment:
There shouldn't even be qdeleted machines in GLOB.machines and even if there are, just skip them instead of canceling the event (which the kill() call doesn't even do correctly).

Reply:
better safe than sorry, if an admin delete the SM crystal or something weird happened i'll just skip ahead since there is a kill() proc later if there is no crystal around

Reply:
Don't check for qdeleted there at all.
If this happens we definitely want things to break.
Instead, make sure it doesn't happen.